Mathlib/RingTheory/DividedPowers/Basic.lean

```diff
@@ -0,0 +1,497 @@
+    {I : Ideal A} (hI : DividedPowers I) (hI' : DividedPowers I)
+    (h_eq : ∀ (n : ℕ) {x : A} (_ : x ∈ I), hI.dpow n x = hI'.dpow n x) :
+    hI = hI' := by
+  obtain ⟨hI, h₀, _⟩ := hI
+  obtain ⟨hI', h₀', _⟩ := hI'
+  simp only [mk.injEq]
+  ext n x
+  by_cases hx : x ∈ I
+  · exact h_eq n hx
+  · rw [h₀ hx, h₀' hx]
+
+theorem DividedPowers.coe_injective {A : Type*} [CommSemiring A] (I : Ideal A) :
+    Function.Injective (fun (h : DividedPowers I) ↦ (h : ℕ → A → A)) := fun hI hI' h ↦ by
+  ext n x
+  exact congr_fun (congr_fun h n) x
+
+end DividedPowersDefinition
+
+namespace DividedPowers
+
+open Finset Nat
+
+section BasicLemmas
+
+/- ## Basic lemmas for divided powers -/
+
+variable {A : Type*} [CommSemiring A] {I : Ideal A}
+
+/-- Variant of `DividedPowers.dpow_add` with a sum on `range (n + 1)` -/
+theorem dpow_add' (hI : DividedPowers I) (n : ℕ) {x y : A} (hx : x ∈ I) (hy : y ∈ I) :
+    hI.dpow n (x + y) = (range (n + 1)).sum fun k => hI.dpow k x * hI.dpow (n - k) y := by
+  rw [hI.dpow_add n hx hy, sum_antidiagonal_eq_sum_range_succ_mk]
+
+/-- The exponential series of an element in the context of divided powers,
+`Σ (dpow n a) X ^ n` -/
+def exp (hI : DividedPowers I) (a : A) :=
+  PowerSeries.mk fun n => hI.dpow n a
+
+theorem exp_add_aux (dp : ℕ → A → A) {a b : A} -- (ha : a ∈ I) (hb : b ∈ I) :
+    (dp_add : ∀ n, dp n (a + b) = (antidiagonal n).sum fun k ↦ dp k.1 a * dp k.2 b) :
+    PowerSeries.mk (fun n ↦ dp n (a + b)) =
+      (PowerSeries.mk fun n ↦ dp n a) * (PowerSeries.mk fun n ↦ dp n b) := by
+  ext n
+  simp only [exp, PowerSeries.coeff_mk, PowerSeries.coeff_mul, dp_add n,
+    sum_antidiagonal_eq_sum_range_succ_mk]
+
+theorem exp_add (hI : DividedPowers I) {a b : A} (ha : a ∈ I) (hb : b ∈ I) :
+    hI.exp (a + b) = hI.exp a * hI.exp b := by
+  ext n
+  simp only [exp, PowerSeries.coeff_mk, PowerSeries.coeff_mul, hI.dpow_add n ha hb]
+
+variable (hI : DividedPowers I)
+
+/- ## Rewriting lemmas -/
+
+theorem dpow_smul (n : ℕ) (a : A) {x : A} (hx : x ∈ I) :
+    hI.dpow n (a • x) = a ^ n • hI.dpow n x := by
+  simp only [smul_eq_mul, hI.dpow_mul, hx]
+
+theorem dpow_mul_right (n : ℕ) {a : A} (ha : a ∈ I) (x : A) :
+    hI.dpow n (a * x) = hI.dpow n a * x ^ n := by
+  rw [mul_comm, hI.dpow_mul n ha, mul_comm]
+
+theorem dpow_smul_right (n : ℕ) {a : A} (ha : a ∈ I) (x : A) :
+    hI.dpow n (a • x) = hI.dpow n a • x ^ n := by
+  rw [smul_eq_mul, hI.dpow_mul_right n ha, smul_eq_mul]
+
+theorem factorial_mul_dpow_eq_pow (n : ℕ) (x : A) (hx : x ∈ I) :
+    (n.factorial : A) * hI.dpow n x = x ^ n := by
+  induction n with
+  | zero => rw [factorial_zero, cast_one, one_mul, pow_zero, hI.dpow_zero hx]
+  | succ n ih =>
+    rw [factorial_succ, mul_comm (n + 1)]
+    nth_rewrite 1 [← (n + 1).choose_one_right]
+    rw [← choose_symm_add, cast_mul, mul_assoc,
+      ← hI.mul_dpow n 1 hx, ← mul_assoc, ih, hI.dpow_one hx, pow_succ, mul_comm]
+
+theorem dpow_eval_zero {n : ℕ} (hn : n ≠ 0) : hI.dpow n 0 = 0 := by
```

Comment:
Have you considered using `NeZero n` here? (not sure if it is a good idea).

Reply:
The issue is that `zero_pow` uses `hn` as an argument and not `[NeZero n]`.